Lib/zipfile/__init__.py

```diff
@@ -1839,7 +1853,27 @@ def _sanitize_windows_name(cls, arcname, pathsep):
         arcname = pathsep.join(x for x in arcname if x)
         return arcname
 
-    def _extract_member(self, member, targetpath, pwd):
+    def _apply_permissions(self, member, path, mode):
+        """
+        Apply ZipFile permissions to a file on the filesystem with
+        specified PreserveMode
+        """
+        if mode == PreserveMode.NONE:
+            return
+
+        # Ignore permissions if the archive was created on Windows
+        if member.create_system == 0:
+            return
+
+        mask = {
+            PreserveMode.SAFE: 0o777,
+            PreserveMode.ALL: 0xFFFF,
+        }
```

Comment:
You could define

```python
class PreserveMode(enum.IntEnum):
    NONE = 0
    SAFE = 0o777
    ALL = 0o7777
```

then here it's just

```python
(member.external_attr >> 16) & mode
```

(However, see my other comment about not restricting this to specific values - users may have arbitrary requirements that an int can capture better.)